packages/sync/src/conflicts/conflictLink.ts

```diff
@@ -1,15 +1,152 @@
+import { ConflictResolutionData } from "./strategies/ConflictResolutionData";
+import { InMemoryCache } from "apollo-cache-inmemory";
+import { getObjectFromCache } from "../offline/storage/defaultStorage";
+import { isMutation } from "../utils/helpers";
+import { BaseStateProvider } from "./base/BaseStateProvider";
+import { ObjectState, ConflictListener, ConflictResolutionStrategies } from ".";
+
+
+/**
+ * Local conflict thrown when data outdates even before sending it to the server.
+ * Can be used to correct any data in flight or shown user another UI to visualize new state
+ */
+export class LocalConflictError extends Error {
+  constructor(private base: any, private variables: any) {
+    super();
+  }
+}
+
+/**
+ * Represents conflict information that was returned from server
+ */
+export interface ConflictInfo {
+  serverState: ConflictResolutionData
+  clientState: ConflictResolutionData
+  // Expected return type of the mutation
+  returnType: string
+}
+
+/**
+ * Configuration for conflict resolution
+ */
+export interface ConflictConfig {
+  /**
+   * Interface that defines how object state is progressed
+   * This interface needs to match state provider supplied on server.
+   */
+  conflictProvider: ObjectState;
+
+  /**
+   * Interface that can be implemented to receive information about the data conflict
+   *
+   * @deprecated see OfflineClient.registerOfflineEventListener
+   */
+  conflictListener?: ConflictListener;
+
+  /**
+   * The conflict resolution strategy your client should use. By default it takes client version.
+   */
+  conflictStrategy?: ConflictResolutionStrategies;
+
+  /**
+   * Base state provider gives ability to save base state
+   */
+  baseState: BaseStateProvider
+}
+
+/**
+ * Conflict handling link implementation that provides ability to determine 
+ */
+export class ConflictLink extends ApolloLink {
+  private link: ApolloLink;
+  private stater: ObjectState;
+  private strategy: ConflictResolutionStrategies | undefined;
+  private listener: ConflictListener | undefined;
+
+  constructor(private config: ConflictConfig, private cache: InMemoryCache) {
+    super();
+    this.link = onError((errorResponse) => {
+      this.conflictErrorHandler(errorResponse);
+    });
+    this.stater = this.config.conflictProvider;
+    this.strategy = this.config.conflictStrategy;
+    this.listener = this.config.conflictListener;
+  }
+
+  public request(
+    operation: Operation,
+    forward: NextLink,
+  ): Observable<FetchResult> | null {
+    if (!isMutation(operation)) {
+      // 🙈 Nothing to do here 
+      return forward(operation);
+    }
+    const currentState = this.stater.currentState(operation.variables);
+    if (!currentState) {
+      // 🙈 Missing state information in request.
+      // No action here
+      return forward(operation);
+    }
+    this.processBaseState(operation, forward);
+    // FIXME removing base state
```

Comment:
@StephenCoady THere are FIXME all over the place to implement things. 
Removing saving/base is not critical for the moment to prove this works and can be done over the time.